Sources/Testing/Events/Recorder/Event.AdvancedConsoleOutputRecorder.swift

```diff
@@ -0,0 +1,826 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2025 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
+//
+
+internal import _TestingInternals
+
+#if os(Windows)
+import WinSDK
+#endif
+
+extension Event {
+  /// A hierarchical console output recorder that renders test results in a tree structure
+  /// after all tests have completed, following strict formatting specifications.
+  ///
+  /// This recorder implements:
+  /// - Post-run buffering with complete tree rendering after final event
+  /// - Unicode box-drawing characters (│, ├─, ╰─) for tree visualization
+  /// - Four distinct line types: Suite Header, Test Case, Issue Sub-Node, Suite Summary
+  /// - 3-space indentation per hierarchy level
+  /// - Right-aligned duration formatting
+  /// - Thread-safe concurrent event handling
+  @_spi(Experimental)
+  public struct AdvancedConsoleOutputRecorder: Sendable {
+    /// Configuration options for the hierarchical console output recorder.
+    public struct Options: Sendable {
+      public var base: Event.ConsoleOutputRecorder.Options
+      public var useHierarchicalOutput: Bool
+      public var showSuccessfulTests: Bool
+      
+      public init() {
+        self.base = Event.ConsoleOutputRecorder.Options()
+        self.useHierarchicalOutput = true
+        self.showSuccessfulTests = true
+      }
+    }
+    
+    /// The options for this recorder.
+    public let options: Options
+    
+    public let write: @Sendable (String) -> Void
+    
+    private struct HierarchyNode: Sendable {
+      let id: Test.ID
+      let name: String
+      let displayName: String?
+      let isSuite: Bool
+      var status: NodeStatus = .running
+      var startTime: Test.Clock.Instant?
+      var endTime: Test.Clock.Instant?
+      var issues: [IssueInfo] = []
+      var children: [String] = [] // Child node IDs
+      var parent: String? // Parent node ID
+      
+      enum NodeStatus: Sendable {
+        case running
+        case passed
+        case failed
+        case skipped
+        case passedWithKnownIssues(count: Int)
+        case passedWithWarnings(count: Int)
+      }
+      
+      struct IssueInfo: Sendable {
+        let issue: Issue
+        let isKnown: Bool
+        let summary: String
+      }
+    }
+    
+    private struct HierarchyContext {
+      var nodes: [String: HierarchyNode] = [:]
+      var rootNodes: [String] = []
+      var runStartTime: Test.Clock.Instant?
+      var runEndTime: Test.Clock.Instant?
+      var isRunCompleted: Bool = false
+      var totalPassed: Int = 0
+      var totalFailed: Int = 0
+      var totalSkipped: Int = 0
+      var totalKnownIssues: Int = 0
+      var totalWarnings: Int = 0
+      
+      var symbolCounts: [String: Int] = [
```

Comment:
Can we use a structure with properties instead of a String-keyed Dictionary for this purpose?